### PR TITLE
fix drush installation of civicrm 4.6

### PR DESCRIPTION
**Pull request by mallezie**
4.6 PR of https://github.com/civicrm/civicrm-core/pull/5601

### CODE REVIEW

**Comment by mallezie** [comment]
I have no idea how this failing test is related. The same PR passes against master https://github.com/civicrm/civicrm-core/pull/5601 so i assume it's a random test-failure.

File: install/civicrm.php [hunk 1]
```diff
@@ -36,6 +36,13 @@ function civicrm_setup($filesDirectory) {
   global $crmPath, $sqlPath, $pkgPath, $tplPath;
   global $compileDir;
 
+  // This is needed to allow CiviCRM to be installed by drush.
+  $pkgPath = $crmPath . DIRECTORY_SEPARATOR . 'packages';
+  set_include_path($crmPath . PATH_SEPARATOR .
+      $pkgPath . PATH_SEPARATOR .
+      get_include_path()
+  );
+
```
Comment (totten) [inline install/civicrm.php]: It's usually preferable to call ```CRM_Core_ClassLoader::singleton()->register();``` to setup autoloading (for core code, for ```packages```, and for ```vendor```). Using the classloader makes it easier to reorganize the codebase (eg shifting classes from ```packages``` to ```vendor```/composer.json; or changing the relationship between the civicrm-core and civicrm-drupal repos).

FWIW, both ```install/index.php``` and ```civicrm.drush.inc``` (```_civicrm_init()``` - used for most commands, albeit perhaps not installation) call ClassLoader's ```register()```.

**Comment by mallezie** [comment]
@totten I suppose this is the suggested way. This also seems to work. 
I added a todo to move this to civicrm-drupal, my tests seems to suggest this also works when added there. But to merge this in civicrm-drupal i would probably need to add a version check.

**Comment by colemanw** [comment]
Not sure if this is confusing you or not, but just to be clear:

Although civicrm-drupal and civicrm-core are separate repos, they are always released syncronously. There is never a situation where anyone would mix-n-match different versions of the two. The Civi release tarball for core 4.6.0 will *always* contain the 7.x-4.6.0 drupal branch, never any other version.

So they do not need to be cross-version-compatible with each other.

**Comment by totten** [comment]
@colemanw, that's generally true, but ```civicrm.drush.inc``` is special. There are a few different use-cases for it:

 1. ```civicrm.drush.inc``` is bundled with the official codebase (via tarball or git). As you say, there's no need for cross-version-compatibility here. However, in this use-case, the drush installer doesn't make any sense -- you download the official tarball, extract it, get a copy of ```civicrm.drush.inc```, which then extracts the official tarball again. That's silly!
 2. ```civicrm.drush.inc``` is used by Aegir/provision_civicrm. In this case, you have one management system which can handle several different codebases (each running different Civi versions), so it's good have to cross-version-compatibility. (The drush installer doesn't make much sense here, either, because Aegir supports drush-make... but maybe provision_civicrm calls some of the internal functions from ```civicrm.drush.inc```?)
 3. ```civicrm.drush.inc``` can be downloaded by itself (directly from Github) and installed in your personal drush folder. Again, you may have several different codebases (each running different Civi versions), so ```civicrm.drush.inc``` should support all of them. But in this use-case, the drush installer *actually makes sense* (because you may have some codebases which don't have a copy of Civi yet).

This is a confusing architecture (and you might say civicrm.drush.inc suffers a chicken-egg problem), but it's what you get when different people need different workflows. It's also part of the subtext to my [installer rewrite rant](http://forum.civicrm.org/index.php/topic,35050.0.html).

**Comment by totten** [comment]
@mallezie , yeah, it might be better to do a version-check+register() in ```civicrm.drush.inc```, but the current core patch does work out to be simpler, and it complies with the class-loading contract, so I'm down with it.

**Comment by colemanw** [comment]
Makes sense, thanks Tim.

**Comment by mallezie** [comment]
Thanks for the explanations. So we leave it at this PR.

**Pull request merged**